lib/ack-tracker/late_ack_record_container_static.c

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright (c) 2002-2019 Balabit
+ * Copyright (c) 2019 Laszlo Budai
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "late_ack_record_container.h"
+#include "ringbuffer.h"
+
+typedef struct _StaticLateAckRecordContainer
+{
+  LateAckRecordContainer super;
+  RingBuffer ack_records;
+  LateAckRecord *pending;
+} StaticLateAckRecordContainer;
+
+static gboolean
+_is_empty(const LateAckRecordContainer *s)
+{
+  StaticLateAckRecordContainer *self = (StaticLateAckRecordContainer *)s;
+
+  return ring_buffer_is_empty(&self->ack_records);
+}
+
+static LateAckRecord *
+_request_pending(LateAckRecordContainer *s)
+{
+  StaticLateAckRecordContainer *self = (StaticLateAckRecordContainer *)s;
+
+  if (self->pending)
+    return self->pending;
+
+  self->pending = ring_buffer_tail(&self->ack_records);
+
+  return self->pending;
+}
+
+static void
+_store_pending(LateAckRecordContainer *s)
+{
+  StaticLateAckRecordContainer *self = (StaticLateAckRecordContainer *)s;
+  g_assert(ring_buffer_push(&self->ack_records) == self->pending);
+  self->pending = NULL;
+}
+
+static void
+_drop(LateAckRecordContainer *s, gsize n)
+{
+  StaticLateAckRecordContainer *self = (StaticLateAckRecordContainer *)s;
+  int i;
+  LateAckRecord *ack_rec;
+  n = MIN(ring_buffer_count(&self->ack_records), n);
```

Comment:
In my understanding it is a bug if `n` is greater than `ring_buffer_count`.
Shouldn't we change this to an assert (`g_assert(n <= ring_buffer_count(&self->ack_records))`)?

Reply:
ok